mayastor/src/subsys/registration/registration_grpc.rs

```diff
@@ -0,0 +1,197 @@
+    hb_interval_sec: Duration,
+    /// how long we wait to send a registration message before timing out
+    hb_timeout_sec: Duration,
+}
+
+/// Registration component for registering dataplane to controlplane
+#[derive(Clone)]
+pub struct Registration {
+    /// Configuration of the registration
+    config: Configuration,
+    /// Registration server grpc endpoint
+    registration_endpoint: tonic::transport::Endpoint,
+    /// Receive channel for messages and termination
+    rcv_chan: async_channel::Receiver<()>,
+    /// Termination channel
+    fini_chan: async_channel::Sender<()>,
+}
+
+static GRPC_REGISTRATION: OnceCell<Registration> = OnceCell::new();
+impl Registration {
+    /// Initialise the global registration instance
+    pub fn init(node: &str, grpc_endpoint: &str, registration_addr: Uri) {
+        GRPC_REGISTRATION.get_or_init(|| {
+            Registration::new(node, grpc_endpoint, registration_addr)
+        });
+    }
+
+    /// Create a new registration instance
+    pub fn new(
+        node: &str,
+        grpc_endpoint: &str,
+        registration_addr: Uri,
+    ) -> Self {
+        let (msg_sender, msg_receiver) = async_channel::unbounded::<()>();
+        let config = Configuration {
+            node: node.to_owned(),
+            grpc_endpoint: grpc_endpoint.to_owned(),
+            hb_interval_sec: match env::var("MAYASTOR_HB_INTERVAL_SEC")
+                .map(|v| v.parse::<u64>())
+            {
+                Ok(Ok(num)) => Duration::from_secs(num),
+                _ => HB_INTERVAL_SEC,
+            },
+            hb_timeout_sec: match env::var("MAYASTOR_HB_TIMEOUT_SEC")
+                .map(|v| v.parse::<u64>())
+            {
+                Ok(Ok(num)) => Duration::from_secs(num),
+                _ => HB_TIMEOUT_SEC,
+            },
+        };
+        let endpoint = tonic::transport::Endpoint::from(registration_addr)
+            .connect_timeout(config.hb_timeout_sec)
+            .timeout(config.hb_timeout_sec);
+        Self {
+            config,
+            registration_endpoint: endpoint,
+            rcv_chan: msg_receiver,
+            fini_chan: msg_sender,
+        }
+    }
+
+    pub(super) fn get() -> Option<&'static Registration> {
+        GRPC_REGISTRATION.get()
+    }
+
+    /// terminate and re-register
+    pub(super) fn fini(&self) {
+        self.fini_chan.close();
+    }
+
+    /// Create a new RegistrationClient for rpc transport
+    pub async fn client(
+        &self,
+    ) -> Result<
+        registration_client::RegistrationClient<tonic::transport::Channel>,
+        ReplyError,
+    > {
+        match registration_client::RegistrationClient::connect(
+            self.registration_endpoint.clone(),
+        )
+        .await
+        {
+            Ok(client) => Ok(client),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    /// Register a new node over rpc
+    pub async fn register(&self) -> Result<(), ReplyError> {
```

Comment:
we should just return tonic::Status here

Reply:
Done.